src/main/java/de/azapps/kafkabackup/common/record/RecordJSONSerde.java

```diff
@@ -43,38 +51,84 @@ public void write(OutputStream outputStream, Record record) throws IOException {
             // TODO: is getting a decoder expensive?
+            String keyBase64 = node.get(KEY_PROPERTY).asText(null); // Default seems to be the string "null", which is not wanted here
             byte[] key = (keyBase64 == null) ? null : Base64.getDecoder().decode(keyBase64);
+            String valueBase64 = node.get(VALUE_PROPERTY).asText(null);
             byte[] value = (valueBase64 == null) ? null : Base64.getDecoder().decode(valueBase64);
 
-            return new Record(topic, partition, key, value, offset, timestamp, timestampType);
+            // Parse headers using registered deserializer
+            ConnectHeaders headers = Optional.ofNullable(node.findValue(HEADERS_PROPERTY))
+                    .map(n -> {
+                        try {
+                            return n.traverse(jp.getCodec()).readValueAs(ConnectHeaders.class);
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                    })
+                    .orElseGet(ConnectHeaders::new);
+
+            return new Record(topic, partition, key, value, offset, timestamp, timestampType, headers);
+        }
+    }
+
+    public static class HeadersDeserializer extends StdDeserializer<ConnectHeaders> {
+        public HeadersDeserializer() {
+            super(ConnectHeaders.class);
+        }
+
+        @Override
+        public ConnectHeaders deserialize(JsonParser jp, DeserializationContext ctx) throws IOException {
+            if (!jp.isExpectedStartArrayToken()) {
+                throw new JsonParseException(jp, "Start array expected");
+            }
+            ConnectHeaders headers = new ConnectHeaders();
+            JsonToken token = jp.nextToken();
+            while (token != JsonToken.END_ARRAY) {
+                headers.add(jp.readValueAs(ConnectHeader.class));
+                token = jp.nextToken();
+            }
+            return headers;
         }
     }
 
-    public static class Serializer extends StdSerializer<Record> {
-        public Serializer() {
+    public static class HeaderDeserializer extends StdDeserializer<ConnectHeader> {
+        public HeaderDeserializer() {
+            super(Header.class);
+        }
+
+        @Override
+        public ConnectHeader deserialize(JsonParser jp, DeserializationContext ctx) throws IOException {
+            JsonNode node = jp.getCodec().readTree(jp);
+            String key = node.get(HEADER_KEY_PROPERTY).asText(null);
+            String valueBase64 = node.get(HEADER_VALUE_PROPERTY).asText(null);
+            byte[] value = (valueBase64 == null) ? null : Base64.getDecoder().decode(valueBase64);
+
+            return new ConnectHeader(key, new SchemaAndValue(Schema.OPTIONAL_BYTES_SCHEMA, value));
+        }
+    }
+
+    public static class RecordSerializer extends StdSerializer<Record> {
+        public RecordSerializer() {
             super(Record.class);
         }
```

Comment:
let's maybe also move those classes to separated files?

Reply:
Yes. Will do in follow-up PR